numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -2142,6 +2144,156 @@ string_inputs_promoter(
+    if (loop_descrs[4] == NULL) {
+        return _NPY_ERROR_OCCURRED_IN_CAST;
+    }
+
+    return NPY_NO_CASTING;
+}
+
+static int
+slice_strided_loop(PyArrayMethod_Context *context, char *const data[],
+                   npy_intp const dimensions[], npy_intp const strides[],
+                   NpyAuxData *NPY_UNUSED(auxdata))
+{
+    char *iptr = data[0];
+    char *start_ptr = data[1];
+    char *stop_ptr = data[2];
+    char *step_ptr = data[3];
+    char *optr = data[4];
+
+    npy_intp N = dimensions[0];
+
+    npy_string_allocator *allocators[5] = {};
+    NpyString_acquire_allocators(5, context->descriptors, allocators);
+    npy_string_allocator *iallocator = allocators[0];
+    npy_string_allocator *oallocator = allocators[4];
+
+    // Build up an index mapping codepoint indices to locations in the encoded
+    // string.
+    std::vector<unsigned char *> codepoint_offsets;
+
+    while (N--) {
+        // get the slice
+        npy_intp start = *(npy_intp *)start_ptr;
+        npy_intp stop = *(npy_intp *)stop_ptr;
+        npy_intp step = *(npy_intp *)step_ptr;
+
+        npy_static_string is = {0, NULL};
+        const npy_packed_static_string *ips = (npy_packed_static_string *)iptr;
+        npy_static_string os = {0, NULL};
+        npy_packed_static_string *ops = (npy_packed_static_string *)optr;
+        int is_isnull = NpyString_load(iallocator, ips, &is);
+        if (is_isnull == -1) {
+            npy_gil_error(PyExc_MemoryError, "Failed to load string in slice");
+            goto fail;
+        }
+        else if (is_isnull) {
+            npy_gil_error(PyExc_TypeError, "Cannot slice null string");
+            goto fail;
+        }
+
+        // number of codepoints in string
+        size_t num_codepoints = 0;
+        // leaves capacity the same as in previous loop iterations to avoid
+        // heap thrashing
+        codepoint_offsets.clear();
+        {
+            const char *inbuf_ptr = is.buf;
+            const char *inbuf_ptr_end = is.buf + is.size;
+
+            // ignore trailing nulls
+            while (inbuf_ptr < inbuf_ptr_end && *(inbuf_ptr_end - 1) == 0) {
+                inbuf_ptr_end--;
+            }
+
+            while (inbuf_ptr < inbuf_ptr_end) {
+                num_codepoints++;
+                int num_bytes = num_bytes_for_utf8_character(
+                        ((unsigned char *)inbuf_ptr));
+                codepoint_offsets.push_back((unsigned char *)inbuf_ptr);
+                inbuf_ptr += num_bytes;
+            }
+        }
+
+        // adjust slice to string length in codepoints
+        // and handle negative indices
+        npy_intp slice_length =
+                PySlice_AdjustIndices(num_codepoints, &start, &stop, step);
+
+        // compute outsize
+        npy_intp outsize = 0;
+        for (int i = start; step > 0 ? i < stop : i > stop; i += step) {
+            outsize += num_bytes_for_utf8_character(codepoint_offsets[i]);
```

Comment:
This seems weird - you already calculated the number of bytes - can one not just take a difference between the relevant offsets? (Alternatively, perhaps there should also be a `std::vector::<unsigned char> code_point_lengths`?)

Reply:
I didn't want to add the extra allocation because this function is so simple it seemed better to just call it twice. That said the allocation is amortized over the array so perhaps it doesn't matter.

What about making `num_bytes_for_utf8_character` a `static inline` function? It's can also make it branchless, I think, with some bitmath...

Reply:
Yes, you're right, it should be possible to count the number of bytes in UTF8 by bitmath; after all, all information on a charcter's length is in the first byte (e.g., http://www.daemonology.net/blog/2008-06-05-faster-utf8-strlen.html)

Reply:
p.s. Totally fine to leave this for a subsequent PR!

Reply:
There's also https://github.com/simdutf/simdutf. But I will leave all of this as an exercise for the reader....

Reply:
Here's a link to the codepoint counting implementation:
https://github.com/simdutf/simdutf/blob/c4c8e0c09c8b65d4d729ae7192f62ae1eac24b4d/src/generic/utf8.h#L10

Reply:
And for AVX512: https://github.com/simdutf/simdutf/blob/c4c8e0c09c8b65d4d729ae7192f62ae1eac24b4d/src/icelake/implementation.cpp#L1272